crypto/aes/aesccm.c

```diff
@@ -0,0 +1,323 @@
+    data++;
+    size--;
+  }
+
+  return EXIT_SUCCESS;
+}
+
+static AES_RETURN cbc_mac_update_zero_padding(cbc_mac_context *ctx) {
+  if (ctx->pos != 0) {
+    if (aes_encrypt(ctx->state, ctx->state, ctx->encrypt_ctx) != EXIT_SUCCESS) {
+      memzero(ctx, sizeof(*ctx));
+      return EXIT_FAILURE;
+    }
+    ctx->pos = 0;
+  }
+  return EXIT_SUCCESS;
+}
+
+static AES_RETURN cbc_mac_final(cbc_mac_context *ctx, uint8_t *mac,
+                                size_t mac_len) {
+  if (ctx->pos != 0 || mac_len > AES_BLOCK_SIZE) {
+    memzero(ctx, sizeof(*ctx));
+    return EXIT_FAILURE;
+  }
+  memcpy(mac, ctx->state, mac_len);
+  memzero(ctx, sizeof(*ctx));
+  return EXIT_SUCCESS;
+}
+
+static AES_RETURN aes_ccm_init(aes_encrypt_ctx *encrypt_ctx,
+                               const uint8_t *nonce, size_t nonce_len,
+                               const uint8_t *adata, size_t adata_len,
+                               size_t plaintext_len, size_t mac_len,
+                               cbc_mac_context *cbc_ctx, uint8_t *ctr_block) {
+  if (mac_len < 4 || mac_len > AES_BLOCK_SIZE || mac_len % 2 != 0) {
+    return EXIT_FAILURE;
+  }
+
+  if (nonce_len < 7 || nonce_len > 13) {
+    return EXIT_FAILURE;
+  }
+
+  // Length of the binary representation of plaintext_len.
+  const size_t q = 15 - nonce_len;
+
+  uint8_t flags = (adata_len != 0) << 6;
+  flags |= ((mac_len - 2) / 2) << 3;
+  flags |= q - 1;
+
+  // Encode the first block.
+  uint8_t block[AES_BLOCK_SIZE] = {0};
+  block[0] = flags;
+  memcpy(&block[1], nonce, nonce_len);
+  size_t shifted_len = plaintext_len;
+  for (size_t i = 0; i < q; ++i) {
+    block[15 - i] = shifted_len & 0xff;
+    shifted_len >>= 8;
+  }
+  if (shifted_len != 0) {
+    // plaintext_len does not fit into q octets.
+    return EXIT_FAILURE;
+  }
+
+  aes_mode_reset(encrypt_ctx);
+  cbc_mac_init(cbc_ctx, encrypt_ctx);
+  if (cbc_mac_update(cbc_ctx, block, sizeof(block)) != EXIT_SUCCESS) {
+    return EXIT_FAILURE;
+  }
+
+  // Format the associated data length.
+  if (adata_len != 0) {
+    size_t block_size = 0;
+    if (adata_len < 65536 - 256) {
+      block[0] = adata_len >> 8;
+      block[1] = adata_len & 0xff;
+      block_size = 2;
+    } else {
+      shifted_len = adata_len;
+      block[5] = shifted_len & 0xff;
+      shifted_len >>= 8;
+      block[4] = shifted_len & 0xff;
+      shifted_len >>= 8;
+      block[3] = shifted_len & 0xff;
+      shifted_len >>= 8;
+      block[2] = shifted_len & 0xff;
+      block[1] = 0xfe;
+      block[0] = 0xff;
+      block_size = 6;
+      if ((shifted_len >> 8) != 0) {
+        // Associated data over 4 GB not supported.
+        return EXIT_FAILURE;
+      }
+    }
+
+    if (cbc_mac_update(cbc_ctx, block, block_size) != EXIT_SUCCESS ||
+        cbc_mac_update(cbc_ctx, adata, adata_len) != EXIT_SUCCESS ||
+        cbc_mac_update_zero_padding(cbc_ctx) != EXIT_SUCCESS) {
+      return EXIT_FAILURE;
+    }
+  }
+
+  // Initialize counter.
+  ctr_block[0] = q - 1;
```

Comment:
Although `ctr_block` was zeroed in the caller's method, I would not rely on that. I would explicitly zero it here.

Reply:
https://github.com/trezor/trezor-firmware/pull/3185/commits/879086fd608dbe569a8309c330be8e562fa6650e